src/core/lib/transport/metadata.h

```diff
@@ -141,19 +145,202 @@ inline bool grpc_mdelem_static_value_eq(grpc_mdelem a, grpc_mdelem b_static) {
    is used as a type tag and is checked during user_data fetch. */
 void* grpc_mdelem_get_user_data(grpc_mdelem md, void (*if_destroy_func)(void*));
 void* grpc_mdelem_set_user_data(grpc_mdelem md, void (*destroy_func)(void*),
-                                void* user_data);
+                                void* data);
+
+// Defined in metadata.cc.
+struct mdtab_shard;
+
+#ifndef NDEBUG
+void grpc_mdelem_trace_ref(void* md, const grpc_slice& key,
+                           const grpc_slice& value, intptr_t refcnt,
+                           const char* file, int line);
+void grpc_mdelem_trace_unref(void* md, const grpc_slice& key,
+                             const grpc_slice& value, intptr_t refcnt,
+                             const char* file, int line);
+#endif
+namespace grpc_core {
+
+typedef void (*destroy_user_data_func)(void* data);
+
+struct UserData {
+  Mutex mu_user_data;
+  grpc_core::Atomic<intptr_t> destroy_user_data;
```

Comment:
Are these stuck at `Atomic<intptr_t>` rather than `Atomic<actual_object_type>` for API reasons or some other reason?

Reply:
The alternative is Atomic<void*> for the user data, and Atomic<destroy_user_data_func> for the destroy_user_data member.

Historically gpr_atm is an intptr_t, so I kept it for commonality. I think it's a bit cleaner looking to leave it as intptr_t; fundamentally it doesn't make a difference in the generated code.

Reply:
Updated per request.